<a href="https://colab.research.google.com/github/Felipe-Bach/carpriceanalysis_ibm/blob/main/Car_Price_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Importing Data

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pylab as plt
from google.colab import files

In [2]:
url = 'http://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data'

In [3]:
df = pd.read_csv(url, encoding='latin-1', header = None)

In [4]:
df.head()

0    1            2    3    4     5   ...    20   21    22  23  24     25
0   3    ?  alfa-romero  gas  std   two  ...   9.0  111  5000  21  27  13495
1   3    ?  alfa-romero  gas  std   two  ...   9.0  111  5000  21  27  16500
2   1    ?  alfa-romero  gas  std   two  ...   9.0  154  5000  19  26  16500
3   2  164         audi  gas  std  four  ...  10.0  102  5500  24  30  13950
4   2  164         audi  gas  std  four  ...   8.0  115  5500  18  22  17450

[5 rows x 26 columns]

In [5]:
headers = ['symboling', 'normalized-losses', 'maker', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length',
          'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm',
           'city-mpg', 'highway-mpg', 'price']                  

In [6]:
df.columns = headers

In [7]:
df.head()

symboling normalized-losses        maker  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [8]:
# df.to_csv('carsprice.csv') 
# files.download('carsprice.csv')

##Analyzing Data

In [12]:
df.dtypes

symboling              int64
normalized-losses     object
maker                 object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

In [13]:
df.describe(include='all')

symboling normalized-losses   maker  ...    city-mpg highway-mpg price
count   205.000000               205     205  ...  205.000000  205.000000   205
unique         NaN                52      22  ...         NaN         NaN   187
top            NaN                 ?  toyota  ...         NaN         NaN     ?
freq           NaN                41      32  ...         NaN         NaN     4
mean      0.834146               NaN     NaN  ...   25.219512   30.751220   NaN
std       1.245307               NaN     NaN  ...    6.542142    6.886443   NaN
min      -2.000000               NaN     NaN  ...   13.000000   16.000000   NaN
25%       0.000000               NaN     NaN  ...   19.000000   25.000000   NaN
50%       1.000000               NaN     NaN  ...   24.000000   30.000000   NaN
75%       2.000000               NaN     NaN  ...   30.000000   34.000000   NaN
max       3.000000               NaN     NaN  ...   49.000000   54.000000   NaN

[11 rows x 26 columns]

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  205 non-null    object 
 2   maker              205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       205 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  engine-size        205 non

##Pre-Processing

Changing price type to float

In [17]:
#df['price'] = df['price'].astype(float)

Error returned because of str in price column ('?')

Cleanning DataFrame from '?' value

In [18]:
df.replace("?", np.nan, inplace = True)
df.head(5)

symboling normalized-losses        maker  ... city-L/100km highway-mpg  price
0          3               NaN  alfa-romero  ...    11.190476          27  13495
1          3               NaN  alfa-romero  ...    11.190476          27  16500
2          1               NaN  alfa-romero  ...    12.368421          26  16500
3          2               164         audi  ...     9.791667          30  13950
4          2               164         audi  ...    13.055556          22  17450

[5 rows x 26 columns]

##Data Wrangling

###Counting missing values per column

In [19]:
missing_data = df.isnull()
missing_data.head(5)

symboling  normalized-losses  maker  ...  city-L/100km  highway-mpg  price
0      False               True  False  ...         False        False  False
1      False               True  False  ...         False        False  False
2      False               True  False  ...         False        False  False
3      False              False  False  ...         False        False  False
4      False              False  False  ...         False        False  False

[5 rows x 26 columns]

In [20]:
for column in missing_data.columns.values.tolist():
    print(column)
    print(missing_data[column].value_counts())
    print("")    

symboling
False    205
Name: symboling, dtype: int64

normalized-losses
False    164
True      41
Name: normalized-losses, dtype: int64

maker
False    205
Name: maker, dtype: int64

fuel-type
False    205
Name: fuel-type, dtype: int64

aspiration
False    205
Name: aspiration, dtype: int64

num-of-doors
False    203
True       2
Name: num-of-doors, dtype: int64

body-style
False    205
Name: body-style, dtype: int64

drive-wheels
False    205
Name: drive-wheels, dtype: int64

engine-location
False    205
Name: engine-location, dtype: int64

wheel-base
False    205
Name: wheel-base, dtype: int64

length
False    205
Name: length, dtype: int64

width
False    205
Name: width, dtype: int64

height
False    205
Name: height, dtype: int64

curb-weight
False    205
Name: curb-weight, dtype: int64

engine-type
False    205
Name: engine-type, dtype: int64

num-of-cylinders
False    205
Name: num-of-cylinders, dtype: int64

engine-size
False    205
Name: engine-size, dtype: int64

fuel-system


Based on the summary above, each column has 205 rows of data and seven of the columns containing missing data:

<ol>
    <li>"normalized-losses": 41 missing data</li>
    <li>"num-of-doors": 2 missing data</li>
    <li>"bore": 4 missing data</li>
    <li>"stroke" : 4 missing data</li>
    <li>"horsepower": 2 missing data</li>
    <li>"peak-rpm": 2 missing data</li>
    <li>"price": 4 missing data</li>
</ol>


Replacing data:
 
By Mean: 'normalized-losses', 'bore', 'stroke', 'horsepower', 'peak-rpm'. 

By Frequency: 'num-of-doors'

Drop row: 'price'. 

### Mean column "Normalized Losses"

In [21]:
mean_normloss = df['normalized-losses'].astype(float).mean(axis=0)
print(f'Mean of normalized losses: {mean_normloss}')

Mean of normalized losses: 122.0


In [22]:
df["normalized-losses"].replace(np.nan, mean_normloss, inplace=True)

###Mean column "Bore"

In [23]:
mean_bore = df['bore'].astype(float).mean(axis=0)
print(f'Mean of bore: {mean_bore}')

Mean of bore: 3.3297512437810957


In [24]:
df['bore'].replace(np.nan, mean_bore, inplace=True)

###Mean column 'Stroke'

In [25]:
mean_stroke = df['stroke'].astype(float).mean(axis=0)
print(f'Mean of column Stroke: {mean_stroke}')

Mean of column Stroke: 3.2554228855721337


In [26]:
df['stroke'].replace(np.nan, mean_stroke, inplace = True)

###Mean column 'Horsepower'

In [27]:
mean_hp = df['horsepower'].astype(float).mean(axis=0)
print(f'mean of column HorsePower: {mean_hp}')

mean of column HorsePower: 104.25615763546799


In [28]:
df['horsepower'].replace(np.nan, mean_hp, inplace=True)

###Mean column "Peak-RPM"

In [29]:
mean_peak_rpm = df['peak-rpm'].astype(float).mean(axis=0)
print(f'Mean of Peak-RPM: {mean_peak_rpm}')

Mean of Peak-RPM: 5125.369458128079


In [30]:
df['peak-rpm'].replace(np.nan, mean_peak_rpm, inplace=True)

###Num of Doors 

Checking most commom type of data on column 'num-of-doors'

In [31]:
df['num-of-doors'].value_counts().idxmax()

'four'

In [32]:
df['num-of-doors'].replace(np.nan, 'four', inplace=True)

###Dropping price column


In [33]:
df.dropna(subset=['price'], axis=0, inplace=True)

Reseting Index

In [34]:
df.reset_index(drop=True, inplace=True)

In [35]:
df.head()

symboling normalized-losses        maker  ... city-L/100km highway-mpg  price
0          3             122.0  alfa-romero  ...    11.190476          27  13495
1          3             122.0  alfa-romero  ...    11.190476          27  16500
2          1             122.0  alfa-romero  ...    12.368421          26  16500
3          2               164         audi  ...     9.791667          30  13950
4          2               164         audi  ...    13.055556          22  17450

[5 rows x 26 columns]

###Converting Data to correct format

In [36]:
df[['bore', 'stroke', 'peak-rpm', 'price']] = df[['bore', 'stroke', 'peak-rpm', 'price']].astype(float) 

In [37]:
df[['normalized-losses']] = df[['normalized-losses']].astype(int)

In [38]:
df.dtypes

symboling              int64
normalized-losses      int64
maker                 object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                 float64
stroke               float64
compression-ratio    float64
horsepower            object
peak-rpm             float64
city-L/100km         float64
highway-mpg            int64
price                float64
dtype: object

###Data Standardization

In [41]:
df.head()

symboling  normalized-losses  ... highway-mpg    price
0          3                122  ...          27  13495.0
1          3                122  ...          27  16500.0
2          1                122  ...          26  16500.0
3          2                164  ...          30  13950.0
4          2                164  ...          22  17450.0

[5 rows x 26 columns]

In [ ]:
df['city-mpg'] = 235/df['city-mpg']
df.rename(columns={'city-mpg':'city-L/100km'}, inplace=True)

In [42]:
df['highway-mpg'] = 235/df['highway-mpg']

df.rename(columns={'highway-mpg': 'highway-L/100km'}, inplace=True)
          
df.head()

symboling  normalized-losses  ... highway-L/100km    price
0          3                122  ...        8.703704  13495.0
1          3                122  ...        8.703704  16500.0
2          1                122  ...        9.038462  16500.0
3          2                164  ...        7.833333  13950.0
4          2                164  ...       10.681818  17450.0

[5 rows x 26 columns]

###Data Normalization 